src/go/rpk/pkg/cli/topic/trim.go

```diff
@@ -0,0 +1,279 @@
+		} else {
+			parsedOffset, err = parseOffsetFlag(ctx, adm, topic, offset, p)
+			if err != nil {
+				return nil, fmt.Errorf("unable to parse offset %v: %v", offset, err)
+			}
+		}
+		o.Add(kadm.Offset{
+			Topic:     topic,
+			Partition: int32(p),
+			At:        parsedOffset,
+		})
+	}
+	return o, nil
+}
+
+// parseOffsetFile parse the given file whether it's a json, yaml or
+// space-delimited text. File content: topic | partition | offset.
+func parseOffsetFile(fs afero.Fs, file string) (kadm.Offsets, error) {
+	parsed, err := out.ParseFileArray[struct {
+		Topic     string `yaml:"topic" json:"topic"`
+		Partition int32  `yaml:"partition" json:"partition"`
+		Offset    int64  `yaml:"offset" json:"offset"`
+	}](fs, file)
+	if err == nil {
+		o := make(kadm.Offsets)
+		for _, element := range parsed {
+			o.Add(kadm.Offset{
+				Topic:     element.Topic,
+				Partition: element.Partition,
+				At:        element.Offset,
+			})
+		}
+		return o, nil
+	}
+	return nil, err
+}
+
+// parseOffsetFlag parse the offset string that we get from the flag, it parses
+// a simple integer or a timestamp-based offset (prefixed with @) it follows the
+// same syntax/decode as 'rpk topic consume --offset' flag.
+func parseOffsetFlag(ctx context.Context, adm *kadm.Client, topic, offset string, partition int) (int64, error) {
+	// Timestamp offset
+	if strings.HasPrefix(offset, "@") {
+		offset = offset[1:] // strip @
+		now := time.Now()
+		_, startAt, end, _, err := parseTimestampBasedOffset(offset, time.Now())
+		if err != nil {
+			return 0, err
+		} else if end {
+			return 0, errors.New("invalid offset '@end'; use '--to-hwm' if you want to trim all")
+		} else if now.Before(startAt) {
+			return 0, fmt.Errorf("invalid offset %q; rpk topic trim does not support a time onward,  use '--to-hwm' if you want to trim all", offset)
+		}
+
+		listed, err := adm.ListOffsetsAfterMilli(ctx, startAt.UnixMilli(), topic)
+		if err == nil {
+			err = listed.Error()
+		}
+		if err != nil {
+			return 0, fmt.Errorf("unable to list offsets after milli %d: %v", startAt.UnixMilli(), err)
+		}
+		filteredOffset, ok := listed.Lookup(topic, int32(partition))
+		if !ok {
+			return 0, fmt.Errorf("unable to find the offset after milli %d in the topic %q and partition %v", startAt.UnixMilli(), topic, partition)
+		}
+		return filteredOffset.Offset, nil
+	}
+	// Not @-prefixed offset
+	return strconv.ParseInt(offset, 10, 64)
+}
+
+// printDeleteRecordRequest prints the passed kadm.Offsets (used for the
+// deleteRecords request). It finds the start offset so the user can see what's
+// the change that is being done, it also finds the hwm for when the user passes
+// the '--to-hwm' flag.
+func printDeleteRecordRequest(ctx context.Context, adm *kadm.Client, o kadm.Offsets) (rerr error) {
+	tw := out.NewTable("TOPIC", "PARTITION", "PRIOR-START-OFFSET", "NEW-START-OFFSET")
+	o.Each(func(offset kadm.Offset) {
+		listed, err := adm.ListStartOffsets(ctx, offset.Topic)
```

Comment:
This needs to be done only once, above the Each.

Reply:
Not really, you can have multiple Topics in kadm.Offsets (e.g: --from-file with a file that describes multiple topics)